File: DIP/Spoilers/MarsRoverKata/Behaviors/Shooter.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace MarsRoverKata.Behaviors
{
    public class Shooter : IBehavior
    {
        public Alien Parent { get; set; }
        private int Speed;
        private readonly Random RNG;

        public Shooter()
        {
            Speed = 3;
            RNG = new Random();
        }

        public void ExecuteBehavior()
        {
            Parent.Facing = pickDirection();

            for (int i = 0; i < Speed; i++)
            {
                //var delta = Target.Location - Parent.Location;

                //var targetDirection = directionToTarget(delta);

                Parent.MoveForward(); // new ForwardCommand(Parent).Execute() : new TurnRightCommand(Parent).Execute();
            }

            Parent.Facing = pickDirection();

            Parent.FireMissle();
        }

        private Direction pickDirection()
        {
            switch (RNG.Next(1, 5))
            {
                case 1:
                    return Direction.North;
                case 2:
                    return Direction.East;
                case 3:
                    return Direction.South;
                case 4:
                    return Direction.West;
            }

            return Direction.North;
        }

    }
}

File: DIP/Java_Final/MarsRoverKata/src/main/java/MarsRoverKata/Behaviors/Shooter.java
package MarsRoverKata.Behaviors;

import MarsRoverKata.Alien;
import MarsRoverKata.Direction;

import java.util.Random;

public class Shooter implements IBehavior {
    private Alien parent;

    public Alien getParent() { return parent; }

    public void setParent(Alien parent) { this.parent = parent; }
    private int Speed;
    private final Random RNG;

    public Shooter()
    {
        Speed = 3;
        RNG = new Random();
    }

    public void executeBehavior()
    {
        getParent().setFacing(pickDirection());

        for (int i = 0; i < Speed; i++)
        {
            //var delta = Target.Location - Parent.Location;

            //var targetDirection = directionToTarget(delta);

            getParent().moveForward(); // new ForwardCommand(Parent).Execute() : new TurnRightCommand(Parent).Execute();
        }

        getParent().setFacing(pickDirection());

        getParent().fireMissile();
    }

    private Direction pickDirection()
    {
        switch (RNG.nextInt(4))
        {
            case 0:
                return Direction.North;
            case 1:
                return Direction.East;
            case 2:
                return Direction.South;
            case 3:
                return Direction.West;
        }

        return Direction.North;
    }
}
